### ISSUE TITLE
Info on sentencepiece with fairseq tokenization (huggingface conversion issues)

### ISSUE THREAD

**Issue by MatejUlcar** [pytorch/fairseq#2572]
## ❓ Questions and Help

I'm training (from scratch) a Roberta-like model, using sentencepiece for tokenization. Basically following Camembert example. 
I've successfully created/trained a sentencepiece bpe model and applied it to the training data. During preprocessing with fairseq, a dict.txt was created (I don't remember the exact command used, I *think* I didn't supply the existing vocab from sp model as a parameter). This new dict.txt has a bit over 80k tokens, while the sentencepiece model has exactly 60k. During the pretraining of roberta model, only the sentencepiece model is provided as a parameter to fairseq train command.

What exactly is the role of each of these two files (sp bpe model and dict.txt)? As I understand, the sp model is just tokenizator, splitting the words into subword tokens, but how do these tokens get fed in the model then? The input layer seems to have dimension of over 80k (exact same size as dict.txt). So I imagine the preprocessed data has that many different tokens/token_ids. I played around a bit with using a partially trained model on a "fill mask" task, and changing sp bpe with another (from a different language model even) only slightly alters the results, I reckon mostly due to worse tokenization. But changing a dict.txt makes it all very wrong. Supplying vocabulary from the sentencepiece model as dict.txt also results in awful performance, only the dict.txt created during preprocessing works.

So, the model as of now seems to perform very well using fairseq and the files as described above. I've been trying to convert it to use it with huggingface's library, but the results are awful. The converted model seems kinda dead-set on only having 60k tokens, and with the order of the two being off, I guess this is where the issue comes from. I'd like to know more about how exactly the tokenization+encoding works, so I can try to tackle the issue if at all possible. I hope someone can help me or at least point me towards the right bits of code, as I don't seem to be able to find anything useful.

**Comment by hichiaty** [comment]
@MatejUlcar I have the same issue, I pre-trained the standard roberta by encoding my dataset (as pieces) with sentencepiece first using [this script](https://github.com/pytorch/fairseq/blob/master/examples/translation/prepare-iwslt17-multilingual.sh) as suggested in the multilingual translation example. I then used fairseq-preprocess with this command on the sentencepiece encoded textfile:
```
fairseq-preprocess \
    --only-source \
    --trainpref predat/encoded_all_spm.txt \
    --validpref predat/eval_all_spm.txt \
    --destdir data-bin\
    --workers 8
```
This generated the dict.txt file after which I simply trained on the data-bin.

I am able to load my roberta model in fairseq and encode the text but unable to use the `fill_mask` method as it expects a base10 int. I have converted to huggingface successfully but have tokenization issues.

Sorry I don't have a solution but thought I'd add the preprocess command I used.
